File: src/lex.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <ctype.h>
#include "lex.h"

Lexer init_lex(const char* source) {
    return (Lexer){
        .source = source,
        .src_len = strlen(source),
        .src_pos = 0,
        .src_line = 1,
        .src_col = 1
    };
}

char peek_char(Lexer* lexer){
    if (lexer->src_pos >= lexer->src_len) return '\0';
    return lexer->source[lexer->src_pos];
}

void advance(Lexer* lexer){
    if (lexer->src_pos >= lexer->src_len)
        return;
    char c = lexer->source[lexer->src_pos++];
    if (c == '\n') {
        lexer->src_line++;
        lexer->src_col = 1;
    } else {
        lexer->src_col++;
    }
}

void skip_nontokens(Lexer* lexer){
    while (1){
        char c = peek_char(lexer);

        if (c == ' ' || c == '\t' || c == '\r') {
            advance(lexer);
        } else {
            break;
        } 
    }
}

Token next_token(Lexer* lexer) {
    skip_nontokens(lexer);

    size_t start = lexer->src_pos;
    int line = lexer->src_line;
    int col = lexer->src_col;

    if (lexer->src_pos >= lexer->src_len)
        return (Token){TOK_EOF, start, 0, line, col};

    char c = peek_char(lexer);

    if (isalpha(c) || c == '_'){
        detect_ident(lexer);
        return (Token){TOK_IDENT, start, lexer->src_pos - start, line, col};
    }

    if ((isdigit(c))){
        detect_number(lexer);
        return (Token){TOK_NUMBER, start,lexer->src_pos - start, lexer->src_line, lexer->src_col};
    }

    switch (c) {
        case '(':
            advance(lexer);
            return (Token){TOK_PARENL, start, 1, line, col};
        case ')':
            advance(lexer);
            return (Token){TOK_PARENR, start, 1, line, col};
        case '+':
            advance(lexer);
            return (Token){TOK_PLUS, start, 1, line, col};
        case '-':
            advance(lexer);
            return (Token){TOK_MINUS, start, 1, line, col};
        case ';':
            while (peek_char(lexer) != '\n' && peek_char(lexer) != '\0')
                advance(lexer);
            return (Token){TOK_COMMENT, start, lexer->src_pos - start, line, col};
        case '"':
            do{
                advance(lexer);
            } while (peek_char(lexer) != '"' && peek_char(lexer) != '\n' && peek_char(lexer) != '\0');
            return (Token){TOK_STRING, start, lexer->src_pos - start, line, col};
        default:
            advance(lexer);
            return (Token){TOK_UNKNOWN, start, 1, line, col};
        }
}

void detect_number(Lexer *lexer) {
    typedef enum {
        START, STARTZERO,
        DEC, DECDOT, DECFRAC, DECEXP, DECEXPSIGN, DECEXPDIG,
        BIN, OCT,
        HEX, HEXDOT, HEXFRAC, HEXEXP, HEXEXPSIGN, HEXEXPDIG,
        END
    } State;

    State state = START;

    while (state != END) {
        char c = peek_char(lexer);
        if (c == '\0') break;

        switch (state) {

        case START:
            if (c == '0') { state = STARTZERO; advance(lexer); }
            else if (isdigit(c)) { state = DEC; advance(lexer); }
            else state = END;
            break;

        case STARTZERO:
            if (c == 'b' || c == 'B') { state = BIN; advance(lexer); }
            else if (c == 'o' || c == 'O') { state = OCT; advance(lexer); }
            else if (c == 'x' || c == 'X') { state = HEX; advance(lexer); }
            else if (isdigit(c)) { state = DEC; advance(lexer); }
            else if (c == '.') { state = DECDOT; advance(lexer); }
            else if (c == 'e' || c == 'E') { state = DECEXP; advance(lexer); }
            else state = END;
            break;

        case DEC:
            if (isdigit(c)) advance(lexer);
            else if (c == '.') { state = DECDOT; advance(lexer); }
            else if (c == 'e' || c == 'E') { state = DECEXP; advance(lexer); }
            else state = END;
            break;

        case DECDOT:
            if (isdigit(c)) { state = DECFRAC; advance(lexer); }
            else state = END;
            break;

        case DECFRAC:
            if (isdigit(c)) advance(lexer);
            else if (c == 'e' || c == 'E') { state = DECEXP; advance(lexer); }
            else state = END;
            break;

        case DECEXP:
            if (c == '+' || c == '-') { state = DECEXPSIGN; advance(lexer); }
            else if (isdigit(c)) { state = DECEXPDIG; advance(lexer); }
            else state = END;
            break;

        case DECEXPSIGN:
            if (isdigit(c)) { state = DECEXPDIG; advance(lexer); }
            else state = END;
            break;

        case DECEXPDIG:
            if (isdigit(c)) advance(lexer);
            else state = END;
            break;

        case BIN:
            if (c == '0' || c == '1') advance(lexer);
            else state = END;
            break;

        case OCT:
            if (c >= '0' && c <= '7') advance(lexer);
            else state = END;
            break;

        case HEX:
            if (isxdigit(c)) advance(lexer);
            else if (c == '.') { state = HEXDOT; advance(lexer); }
            else if (c == 'p' || c == 'P') { state = HEXEXP; advance(lexer); }
            else state = END;
            break;

        case HEXDOT:
            if (isxdigit(c)) { state = HEXFRAC; advance(lexer); }
            else state = END;
            break;

        case HEXFRAC:
            if (isxdigit(c)) advance(lexer);
            else if (c == 'p' || c == 'P') { state = HEXEXP; advance(lexer); }
            else state = END;
            break;

        case HEXEXP:
            if (c == '+' || c == '-') { state = HEXEXPSIGN; advance(lexer); }
            else if (isdigit(c)) { state = HEXEXPDIG; advance(lexer); }
            else state = END;
            break;

        case HEXEXPSIGN:
            if (isdigit(c)) { state = HEXEXPDIG; advance(lexer); }
            else state = END;
            break;

        case HEXEXPDIG:
            if (isdigit(c)) advance(lexer);
            else state = END;
            break;

        case END:
            advance(lexer); return;
        }
    }
}

void detect_ident(Lexer *lexer){
    while (isalnum(peek_char(lexer)) || peek_char(lexer) == '_'){
        advance(lexer);
    }
}

void init_tokstream(TokenStream* ts){
    ts->len = 0;
    ts->cap = 16;
    ts->data = malloc(ts->cap * sizeof(Token));
}

void push_tokstream(TokenStream* ts, Token tok){
    if (ts->len >= ts->cap){
        ts->cap *= 2;
        ts->data = realloc(ts->data, ts->cap * sizeof(Token));
        if (!ts->data)
            exit(1);
    }
    ts->data[ts->len++] = tok;
}

void free_tokstream(TokenStream* ts){
    free(ts->data);
    ts->data = NULL;
    ts->len = 0;
    ts->cap = 0;
}

File: src/lex.rs
#[derive(Debug, Default)]
pub struct Token{
    pub token_type : TokenType,
    pub src_line : usize,
    pub src_col : usize,
    pub tok_pos : usize,
    pub tok_len : usize
}

#[derive(Debug, Default)]
pub enum TokenType{
    EOF,
    Ident,
    Keyword(Keyword),
    Literal(LiteralType),
    Punc(Punc),
    Attribute,              // unused
    #[default]
    Error
}

#[derive(Debug)]
pub enum Keyword{
    // Data Types
    ISIZE, I8, I16, I32, I64, I128,
    USIZE, U8, U16, U32, U64, U128,
    FIXSIZE, FIX8, FIX16, FIX32, FIX64, FIX128,
    FLT16, FLT32, FLT64, FLT128, BFLT16,
    BOOL,

    // Literals
    NULL,
    TRUE, FALSE,

    // Control FLow
    IF, ELSE, MATCH,
    LOOP, FOR, WHILE,
    BREAK, SKIP,
    RETURN,
    DEFER,

    // Contracts
    PRE, INV, POST,

    // Aggregate Data
    STRUCT, UNION, ENUM, SUM,

    // Top-Level Definitions
    VAR, FN, USE,

    // Intermediate Clauses
    AS, TO, IN, OF,

    // Qualifiers
    CONST, PUB, VOL, PACK,

    // Testing
    TEST, ASSERT,

    // Coroutines
    COROUT, RUN, YIELD, CANCEL,

    // Interop
    ASM, EXPORT, EXTERN,

    // Casts
    BITCAST, VALCAST,

    // Type Introspection and Alignment
    ALIGN, SIZE, TYPE,

    // Bitwise and Boolean Operators
    AND, OR, NOT,
    XOR, SHL, SHR,
    EQ, NEQ, GT, GTE, LT, LTE
}

#[derive(Debug)]
pub enum Punc{
    // Non-Alphabetical Operators
    Assignment,                             // =
    Plus, Minus, Mul, Div, Mod,             // + - * / %
    Dot, Arrow, FatArrow, Range, RangeIncl, // . -> => .. ..=
    Wildcard,                               // _
    Quantifier,                             // :
    Pointer, Address,                       // ^ @
    BrackL, BrackR, ParenL, ParenR,          // [] ()

    // Syntax-Related Punctuators
    BraceL, BraceR,         // {}
    Comma, Semicolon,       // , ;
    GenL, GenR,             // <>
    Doublequote, Dollar,    // " $
    Docs,                    // ##, not relevant for actual codegen
    Comment,
    Unknown
}

#[derive(Debug)]
pub enum LiteralType{
    Int,
    Real,
    Char,
    String
}

#[derive(Debug)]
pub struct Lexer<'src>{
    source: &'src[u8],
    pub length: usize,
    pub pos: usize,
    line: usize,
    col: usize
}

static CHAR_CLASS: [u8; 128] = {
    let mut table = [0; 128];
    let mut i  = 0;
    while i < 128 {
        table[i] = match i as u8{
            b'a' ..= b'z' | b'A' ..= b'Z' => 1,
            b'0' ..= b'9' => 2,
            b'=' | b'+' | b'-' | b'*' | b'/' | b'%' | b'.' | b'_' | b':' | b'^' | b'@' |
            b'[' | b']' | b'(' | b')' | b'{' | b'}' | b',' | b';' | b'<' | b'>' | b'"' |
            b'#' | b'$' => 3,
            _ => 0
        };
        i += 1;
    }
    table
};

impl<'src> Lexer<'src>{
    pub fn new(source: &'src str) -> Self{
        let mut bytes = Vec::with_capacity(source.len() + 1);
        bytes.extend_from_slice((source.as_bytes()));
        bytes.push(0);
        let slice = Box::leak(bytes.into_boxed_slice());
        Lexer{
            source: slice,
            length: source.len() + 1,
            pos: 0,
            line: 1,
            col: 1
        }
    }

    fn peek(&self) -> *const u8{
        &self.source[self.pos]
    }

    fn advance(&mut self){
        unsafe{
            match *self.peek(){
                0 => return,
                b'\n' => {
                    self.line += 1;
                    self.col = 1;
                },
                _ => self.col += 1,
            }
        }
        self.pos += 1;
    }

    fn skip_nontokens(&mut self) {
        loop {
            unsafe {
                match *self.peek() {
                    b' ' | b'\t' | b'\r' => self.advance(),
                    _ => break,
                }
            }
        }
    }

    pub fn create_token(&mut self) -> Token {
        self.skip_nontokens();
        let start = self.pos;

        unsafe{
            match CHAR_CLASS[*self.peek() as usize]{
                1 => {
                    return Token { 
                        token_type : self.detect_id_or_kwrd(),
                        src_line : self.line,
                        src_col : self.col,
                        tok_pos : start,
                        tok_len : self.pos - start
                    }       
                },
                2 => {
                    return Token { 
                        token_type : TokenType::Literal(self.detect_number()),
                        src_line : self.line,
                        src_col : self.col,
                        tok_pos : start,
                        tok_len : self.pos - start
                    }
                },
                3 => {
                    return Token { 
                        token_type : TokenType::Punc(self.detect_punc()),
                        src_line : self.line,
                        src_col : self.col,
                        tok_pos : start,
                        tok_len : self.pos - start
                    }
                },
                _ => {
                    self.advance();
                    if *self.peek() == 0{
                        return Token{
                            token_type : TokenType::EOF,
                            src_line : self.line,
                            src_col : self.col,
                            tok_pos : start,
                            tok_len : self.pos - start
                        }
                    } else {
                        return Token{
                            token_type : TokenType::Error,
                            src_line : self.line,
                            src_col : self.col,
                            tok_pos : start,
                            tok_len : self.pos - start
                        }
                    }
                }
            }
        }
    }

    fn detect_id_or_kwrd(&mut self) -> TokenType{
        let start = self.pos;
        unsafe{
            while (CHAR_CLASS[*self.peek() as usize] == 1)
            || (CHAR_CLASS[*self.peek() as usize] == 2)
            || (*self.peek() == b'_'){
                self.advance();
            }
        }

        if (self.pos - start) > 7 {
            return TokenType::Ident;
        }

        match &self.source[start..self.pos] as &[u8] {
            b"isize" => return TokenType::Keyword(Keyword::ISIZE),
            b"i8" => return TokenType::Keyword(Keyword::I8),
            b"i16" => return TokenType::Keyword(Keyword::I16),
            b"i32" => return TokenType::Keyword(Keyword::I32),
            b"i64" => return TokenType::Keyword(Keyword::I64),
            b"i128" => return TokenType::Keyword(Keyword::I128),
            b"usize" => return TokenType::Keyword(Keyword::USIZE),
            b"u8" => return TokenType::Keyword(Keyword::U8),
            b"u16" => return TokenType::Keyword(Keyword::U16),
            b"u32" => return TokenType::Keyword(Keyword::U32),
            b"u64" => return TokenType::Keyword(Keyword::U64),
            b"u128" => return TokenType::Keyword(Keyword::U128),
            b"fixsize" => return TokenType::Keyword(Keyword::FIXSIZE),
            b"fix8" => return TokenType::Keyword(Keyword::FIX8),
            b"fix16" => return TokenType::Keyword(Keyword::FIX16),
            b"fix32" => return TokenType::Keyword(Keyword::FIX32),
            b"fix64" => return TokenType::Keyword(Keyword::FIX64),
            b"fix128" => return TokenType::Keyword(Keyword::FIX128),
            b"flt16" => return TokenType::Keyword(Keyword::FLT16),
            b"flt32" => return TokenType::Keyword(Keyword::FLT32),
            b"flt64" => return TokenType::Keyword(Keyword::FLT64),
            b"flt128" => return TokenType::Keyword(Keyword::FLT128),
            b"bflt16" => return TokenType::Keyword(Keyword::BFLT16),
            b"bool" => return TokenType::Keyword(Keyword::BOOL),
            b"null" => return TokenType::Keyword(Keyword::NULL),
            b"true" => return TokenType::Keyword(Keyword::TRUE),
            b"false" => return TokenType::Keyword(Keyword::FALSE),
            b"if" => return TokenType::Keyword(Keyword::IF),
            b"else" => return TokenType::Keyword(Keyword::ELSE),
            b"match" => return TokenType::Keyword(Keyword::MATCH),
            b"loop" => return TokenType::Keyword(Keyword::LOOP),
            b"for" => return TokenType::Keyword(Keyword::FOR),
            b"while" => return TokenType::Keyword(Keyword::WHILE),
            b"break" => return TokenType::Keyword(Keyword::BREAK),
            b"skip" => return TokenType::Keyword(Keyword::SKIP),
            b"return" => return TokenType::Keyword(Keyword::RETURN),
            b"defer" => return TokenType::Keyword(Keyword::DEFER),
            b"pre" => return TokenType::Keyword(Keyword::PRE),
            b"inv" => return TokenType::Keyword(Keyword::INV),
            b"post" => return TokenType::Keyword(Keyword::POST),
            b"struct" => return TokenType::Keyword(Keyword::STRUCT),
            b"union" => return TokenType::Keyword(Keyword::UNION),
            b"enum" => return TokenType::Keyword(Keyword::ENUM),
            b"sum" => return TokenType::Keyword(Keyword::SUM),
            b"var" => return TokenType::Keyword(Keyword::VAR),
            b"fn" => return TokenType::Keyword(Keyword::FN),
            b"use" => return TokenType::Keyword(Keyword::USE),
            b"as" => return TokenType::Keyword(Keyword::AS),
            b"to" => return TokenType::Keyword(Keyword::TO),
            b"in" => return TokenType::Keyword(Keyword::IN),
            b"of" => return TokenType::Keyword(Keyword::OF),
            b"const" => return TokenType::Keyword(Keyword::CONST),
            b"pub" => return TokenType::Keyword(Keyword::PUB),
            b"vol" => return TokenType::Keyword(Keyword::VOL),
            b"pack" => return TokenType::Keyword(Keyword::PACK),
            b"test" => return TokenType::Keyword(Keyword::TEST),
            b"assert" => return TokenType::Keyword(Keyword::ASSERT),
            b"corout" => return TokenType::Keyword(Keyword::COROUT),
            b"run" => return TokenType::Keyword(Keyword::RUN),
            b"yield" => return TokenType::Keyword(Keyword::YIELD),
            b"cancel" => return TokenType::Keyword(Keyword::CANCEL),
            b"asm" => return TokenType::Keyword(Keyword::ASM),
            b"export" => return TokenType::Keyword(Keyword::EXPORT),
            b"extern" => return TokenType::Keyword(Keyword::EXTERN),
            b"bitcast" => return TokenType::Keyword(Keyword::BITCAST),
            b"valcast" => return TokenType::Keyword(Keyword::VALCAST),
            b"align" => return TokenType::Keyword(Keyword::ALIGN),
            b"size" => return TokenType::Keyword(Keyword::SIZE),
            b"type" => return TokenType::Keyword(Keyword::TYPE),
            b"and" => return TokenType::Keyword(Keyword::AND),
            b"or" => return TokenType::Keyword(Keyword::OR),
            b"not" => return TokenType::Keyword(Keyword::NOT),
            b"xor" => return TokenType::Keyword(Keyword::XOR),
            b"shl" => return TokenType::Keyword(Keyword::SHL),
            b"shr" => return TokenType::Keyword(Keyword::SHR),
            b"eq" => return TokenType::Keyword(Keyword::EQ),
            b"neq" => return TokenType::Keyword(Keyword::NEQ),
            b"gt" => return TokenType::Keyword(Keyword::GT),
            b"gte" => return TokenType::Keyword(Keyword::GTE),
            b"lt" => return TokenType::Keyword(Keyword::LT),
            b"lte" => return TokenType::Keyword(Keyword::LTE),
            _ => return TokenType::Ident
        }
    }

    fn detect_number(&mut self) -> LiteralType{
        enum State{
            Start, Startzero,
            Dec, DecDot, DecFrac, DecExp, DecExpSign, DecExpDig,
            Bin, Oct,
            Hex, HexDot, HexFrac, HexExp, HexExpSign, HexExpDig
        }

        let mut state = State::Start;

        loop{
            unsafe {
                match state {

                    State::Start => {
                        if *self.peek() == b'0' {
                            state = State::Startzero;
                        } else if CHAR_CLASS[*self.peek() as usize] == 2 {
                            state = State::Dec;
                        } else {
                            return LiteralType::Int;
                        }
                    },

                    State::Startzero => {
                        match *self.peek(){
                            b'b' | b'B' => state = State::Bin,
                            b'o' | b'O' => state = State::Oct,
                            b'x' | b'X' => state = State::Hex,
                            b'.' => state = State::DecDot,
                            b'e' | b'E' => state = State::DecExp,
                            _ => {
                                if CHAR_CLASS[*self.peek() as usize] == 2 {
                                    state = State::Dec;
                                } else {
                                    return LiteralType::Int;
                                }
                            }
                        }
                    },

                    State::Dec => {
                        match *self.peek(){
                            b'.' => state = State::DecDot,
                            b'e' | b'E' => state = State::DecExp,
                            _ => {
                                if CHAR_CLASS[*self.peek() as usize] == 2 {
                                    {};
                                } else {
                                    return LiteralType::Int;
                                }                                
                            }
                        }
                    },
                    
                    State::DecDot => {
                        if CHAR_CLASS[*self.peek() as usize] == 2 {
                            state = State::DecFrac;
                        } else {
                            return LiteralType::Real;
                        }
                    },

                    State::DecFrac => {
                        match *self.peek(){
                            b'e' | b'E' => state = State::DecExp,
                            _ => {
                                if CHAR_CLASS[*self.peek() as usize] == 2 {
                                    {};
                                } else {
                                    return LiteralType::Real;
                                }
                            }
                        }
                    },

                    State::DecExp => {
                        match *self.peek(){
                            b'+' | b'-' => state = State::DecExpSign,
                            _ => {
                                if CHAR_CLASS[*self.peek() as usize] == 2 {
                                    state = State::DecExpDig;
                                } else {
                                    return LiteralType::Real;
                                }
                            }
                        }
                    },

                    State::DecExpSign => {
                        if CHAR_CLASS[*self.peek() as usize] == 2 {
                            state = State::DecExpDig;
                        } else {
                            return LiteralType::Real;
                        }
                    },

                    State::DecExpDig => {
                        if CHAR_CLASS[*self.peek() as usize] == 2 {
                            {};
                        } else {
                            return LiteralType::Real;
                        }
                    },

                    State::Bin => {
                        if (*self.peek() != b'0') && (*self.peek() != b'1'){
                            return LiteralType::Int;
                        }
                    },

                    State::Oct => {
                        if (CHAR_CLASS[*self.peek() as usize] == 2) &&
                        (*self.peek() != b'8') && (*self.peek() != b'9'){
                            {};
                        } else {
                            return LiteralType::Int;
                        }
                    },

                    State::Hex => {
                        match *self.peek(){
                            b'p' | b'P' => state = State::HexExp,
                            b'.' => state = State::HexDot,
                            b'a' ..= b'f' | b'A' ..= b'F' => {},
                            _ => {
                                if CHAR_CLASS[*self.peek() as usize] == 2 {
                                    {};
                                } else {
                                    return LiteralType::Int;
                                }
                            },
                        }
                    },

                    State::HexDot => {
                        match *self.peek(){
                            b'a' ..= b'f' | b'A' ..= b'F' => state = State::HexFrac,
                            _ => {
                                if CHAR_CLASS[*self.peek() as usize] == 2 {
                                    state = State::HexFrac;
                                } else {
                                    return LiteralType::Real;
                                }
                            },
                        }
                    },

                    State::HexFrac => {
                         match *self.peek(){
                            b'p' | b'P' => state = State::HexExp,
                            b'a' ..= b'f' | b'A' ..= b'F' => {},
                            _ => {
                                if CHAR_CLASS[*self.peek() as usize] == 2 {
                                    {};
                                } else {
                                    return LiteralType::Real;
                                }
                            },
                        }
                    },

                    State::HexExp => {
                        match *self.peek(){
                            b'+' | b'-' => state = State::HexExpSign,
                            b'a' ..= b'f' | b'A' ..= b'F' => state = State::HexExpDig,
                            _ => {
                                if CHAR_CLASS[*self.peek() as usize] == 2 {
                                    state = State::HexExpDig;
                                } else {
                                    return LiteralType::Real;
                                }
                            }
                        }
                    },

                    State::HexExpSign => {
                        match *self.peek(){
                            b'a' ..= b'f' | b'A' ..= b'F' => state = State::HexExpDig,
                            _ => {
                                if CHAR_CLASS[*self.peek() as usize] == 2 {
                                    state = State::HexExpDig;
                                } else {
                                    return LiteralType::Real;
                                }
                            }
                        }
                    },

                    State::HexExpDig => {
                        match *self.peek(){
                            b'a' ..= b'f' | b'A' ..= b'F' => {},
                            _ => {
                                if CHAR_CLASS[*self.peek() as usize] == 2 {
                                    {};
                                } else {
                                    return LiteralType::Real;
                                }
                            }
                        }
                    }
                }
            }
            self.advance();
        }
    }

    fn detect_punc(&mut self) -> Punc {
        unsafe{
            let c = *self.peek();
            self.advance();
            match c{
                b'=' => {
                    if *self.peek() == b'>' {
                        self.advance();
                        return Punc::FatArrow;
                    } else {
                        return Punc::Assignment;
                    }
                }
                b'+' => return Punc::Plus,
                b'-' => {
                    if *self.peek() == b'>' {
                        self.advance();
                        return Punc::Arrow;
                    } else {
                        return Punc::Minus;
                    }
                }
                b'*' => return Punc::Mul,
                b'/' => return Punc::Div,
                b'%' => return Punc::Mod,
                b'.' => {
                    if *self.peek() == b'.' {
                        self.advance();
                        if *self.peek() == b'=' {
                            self.advance();
                            return Punc::RangeIncl;
                        } else {
                            return Punc::Range;
                        }
                    } else {
                        return Punc::Dot;
                    }
                }
                b'_' => return Punc::Wildcard,
                b':' => return Punc::Quantifier,
                b'^' => return Punc::Pointer,
                b'@' => return Punc::Address,
                b'[' => return Punc::BrackL,
                b']' => return Punc::BrackR,
                b'(' => return Punc::ParenL,
                b')' => return Punc::ParenR,
                b'{' => return Punc::BraceL,
                b'}' => return Punc::BraceR,
                b',' => return Punc::Comma,
                b';' => return Punc::Semicolon,
                b'<' => return Punc::GenL,
                b'>' => return Punc::GenR,
                b'"' => return Punc::Doublequote,   // Becomes string detector
                b'$' => return Punc::Dollar,        // Becomes string directive
                b'#' => {
                    if *self.peek() == b'#' {
                        self.advance();
                        while *self.peek() != b'\n' && *self.peek() != 0 {
                            self.advance();
                        }
                        return Punc::Docs;
                    } else {
                        while *self.peek() != b'\n' && *self.peek() != 0 {
                            self.advance();
                        }
                        return Punc::Comment;
                    }
                },
                _ => return Punc::Unknown,
            }
        }
    }
}